src/crimson/os/seastore/seastore_types.h

```diff
@@ -1049,45 +1081,142 @@ class laddr_t {
 
   friend auto operator<=>(const laddr_t&, const laddr_t&) = default;
 
-  friend laddr_t operator+(const laddr_t &laddr, const Int &i) {
-    return laddr_t{laddr.value + i};
-  }
-
-  friend laddr_t operator+(const Int &i, const laddr_t &laddr) {
-    return laddr_t{laddr.value + i};
-  }
-
-  friend laddr_t operator-(const laddr_t &laddr, const Int &i) {
-    return laddr_t{laddr.value - i};
-  }
-
+  // the difference between two laddrs usually works as byte distance,
+  // assume the result won't overflow and convert it intto the byte
+  // addressing result automatically.
   friend Int operator-(const laddr_t &l, const laddr_t &r) {
-    return l.value - r.value;
-  }
-
-  laddr_t &operator+=(std::integral auto i) {
-    value += i;
-    return *this;
-  }
-
-  laddr_t &operator-=(std::integral auto i) {
-    value -= i;
-    return *this;
+    auto diff = l.value - r.value;
+    assert(diff <= (std::numeric_limits<Int>::max() >> UNIT_SHIFT));
+    return diff << UNIT_SHIFT;
   }
 
+  friend struct loffset_t;
   friend struct laddr_le_t;
   friend struct pladdr_le_t;
 
 private:
+  // Prevent direct construction of laddr_t with an integer,
+  // always use laddr_t::from_raw_int instead.
+  constexpr explicit laddr_t(Int value) : value(value) {}
+
   Int value;
 };
 
 constexpr laddr_t::Int L_RAW_INT_MAX = std::numeric_limits<laddr_t::Int>::max();
-constexpr laddr_t L_ADDR_MAX = laddr_t(L_RAW_INT_MAX);
+constexpr laddr_t L_ADDR_MAX = laddr_t::from_raw_int(L_RAW_INT_MAX);
 constexpr laddr_t L_ADDR_MIN = laddr_t();
 constexpr laddr_t L_ADDR_NULL = L_ADDR_MAX;
-constexpr laddr_t L_ADDR_ROOT = laddr_t(L_RAW_INT_MAX - 1);
-constexpr laddr_t L_ADDR_LBAT = laddr_t(L_RAW_INT_MAX - 2);
+constexpr laddr_t L_ADDR_ROOT = laddr_t::from_raw_int(L_RAW_INT_MAX - 1);
+constexpr laddr_t L_ADDR_LBAT = laddr_t::from_raw_int(L_RAW_INT_MAX - 2);
+
+struct loffset_t {
```

Comment:
Suggest to introduce a `using loffset_t = uint64_t` to directly replace the original raw uint64_t laddr_t, and a `struct laddr_offset_t` meaning it is a combined structure.

Reply:
Elaborated in https://github.com/ceph/ceph/pull/59182#issuecomment-2295576836